zookeeper/core/component.py

```diff
@@ -69,305 +69,309 @@ class C:
 
-    return __component_delattr__
 
+def __component_str__(instance):
+    field_strings = [
+        _field_key_val_str(
+            field_name, getattr(instance, field_name), color=True, single_line=False
+        )
+        for field_name in instance.__component_fields__.keys()
+    ]
 
-##################################
-# Pretty string representations. #
-##################################
+    joined_str = f",\n{INDENT}".join(field_strings)
 
+    return f"{instance.__class__.__name__}(\n{INDENT}{joined_str}\n)"
 
-# Indent for nesting in the string representation
-INDENT = " " * 4
 
+##########################
+# A component `__init__` #
+##########################
 
-def str_key_val(key, value, color=True, single_line=False):
-    if is_component_class(value.__class__):
-        if single_line:
-            value = repr(value)
-        else:
-            value = f"\n{INDENT}".join(str(value).split("\n"))
-    elif callable(value):
-        value = "<callable>"
-    elif type(value) == str:
-        value = f'"{value}"'
-    return f"{BLUE}{key}{RESET}={YELLOW}{value}{RESET}" if color else f"{key}={value}"
 
+def __component_init__(instance, **kwargs):
+    """
+    Accepts keyword-arguments corresponding to fields defined on the component.
+    """
 
-def __component_repr__(instance):
-    fields = ", ".join(
-        [
-            str_key_val(
-                field_name, getattr(instance, field_name), color=False, single_line=True
+    # Use the `kwargs` to set field values.
+    for name, value in kwargs.items():
+        if name not in instance.__component_fields__:
+            raise TypeError(
+                "Keyword arguments passed to component `__init__` must correspond to "
+                f"component fields. Received non-matching argument '{name}'."
             )
-            for field_name in sorted(instance.__component_fields__)
-        ]
-    )
-    return f"{instance.__class__.__name__}({fields})"
 
+    # Save a shallow-clone of the arguments.
+    instance.__component_instantiated_field_values__ = {**kwargs}
 
-def __component_str__(instance):
-    fields = f",\n{INDENT}".join(
-        [
-            str_key_val(field_name, getattr(instance, field_name))
-            for field_name in sorted(instance.__component_fields__)
-        ]
-    )
-    return f"{instance.__class__.__name__}(\n{INDENT}{fields}\n)"
+    # This will contain configured field values that apply to this instance, if
+    # any, which override everything else.
+    instance.__component_configured_field_values__ = {}
+
+    # This will contain the names of every field of this instance and all
+    # component ancestors for which a value is defined. More names will be added
+    # during configuration.
+    instance.__component_fields_with_values_in_scope__ = set(
+        field.name
+        for field in instance.__component_fields__.values()
+        if field.has_default
+    ) | set(name for name in kwargs.keys())
```

Comment:
This should work as well:
```suggestion
    ) | set(kwargs)
```